### PR TITLE
Add CBOR golden tests

### LINKED ISSUE #5390: Add positive `duplicateCertsTx` tests to `Shelley` and `Conway`
I think it would be an awesome addition if we moved this example to shelley era and added it as a positive test, where:
* In Shelley till Babbage we successfully decode such transaction, but only one certificate is preserved
* In Conway only one certificate is preserved

_Originally posted by @lehins in https://github.com/IntersectMBO/cardano-ledger/pull/5370#discussion_r2487825834_

### PR DESCRIPTION
**Pull request by Soupstraw**
# Description

This PR adds a bunch of CBOR golden tests. There are two kinds of new tests in this PR:
 * Tests that check that empty set decoding in `TxWits` fails starting with `Conway`
 * Tests that check that duplicate `TxCert`s are not decoded starting with `Conway` (close https://github.com/IntersectMBO/cardano-ledger/issues/5390)

I also added `forEachEraVersion` that runs tests with each valid protocol version in that era.

# Checklist

- [x] Commits in meaningful sequence and with useful messages.
- [x] Tests added or updated when needed.
- [x] `CHANGELOG.md` files updated for packages with externally visible changes.  
      **NOTE: _New section is never added with the code changes._** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#changelogmd)).
- [x] Versions updated in `.cabal` and `CHANGELOG.md` files when necessary, according to the
      [versioning process](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [x] Version bounds in `.cabal` files updated when necessary.  
      **NOTE: _If bounds change in a cabal file, that package itself must have a version increase._** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process)).
- [x] Code formatted (use `scripts/fourmolize.sh`).
- [x] Cabal files formatted (use `scripts/cabal-format.sh`).
- [x] CDDL files are up to date (use `scripts/gen-cddl.sh`)
- [x] [`hie.yaml`](https://github.com/intersectmbo/cardano-ledger/blob/master/hie.yaml) updated (use `scripts/gen-hie.sh`).
- [x] Self-reviewed the diff.

### CODE REVIEW

File: libs/cardano-ledger-core/testlib/Test/Cardano/Ledger/Core/Binary/Golden.hs [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+  (ToExpr a, DecCBOR (Annotator a), HasCallStack) =>
+  (a -> a -> Bool) ->
+  Version ->
+  Enc ->
+  a ->
+  Expectation
+expectDecoderSuccessAnnWith equals version enc expected =
+  case decodeEnc @a version enc of
+    Right x | x `equals` expected -> pure ()
+    decResult -> expectationFailure $ diffExprString (Right expected) decResult
```
Comment (lehins) [inline libs/cardano-ledger-core/testlib/Test/Cardano/Ledger/Core/Binary/Golden.hs]: This will result in a better test failure message:
```suggestion
    Left err -> expectationFailure $ "Unexpected decoder failure: " <> show err
    Right x | x `equals` expected -> pure ()
    Right result -> expectationFailure $ diffExprString expected result
```

File: libs/cardano-ledger-core/testlib/Test/Cardano/Ledger/Imp/Common.hs [hunk 2]
```diff
@@ -221,3 +229,7 @@ expectNothingExpr (Just x) =
   assertFailure $
     "Expected Nothing, got Just:\n" <> showExpr x
 expectNothingExpr Nothing = pure ()
+
+forEachEraVersion :: forall era. EraTest era => (Version -> Spec) -> Spec
```
Comment (lehins) [inline libs/cardano-ledger-core/testlib/Test/Cardano/Ledger/Imp/Common.hs]: `EraTest` is unnecessary and `HasCallStack` is missing.
Also this function should go into `Test.Cardano.Ledger.Common`, since it is not specific to imp tests
```suggestion
forEachEraVersion :: forall era. (Era era, HasCallStack) => (Version -> Spec) -> Spec
```

File: libs/cardano-ledger-core/testlib/Test/Cardano/Ledger/Core/Binary/Golden.hs [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+expectDecoderSuccessAnn ::
+  (ToExpr a, DecCBOR (Annotator a), Eq a) => Version -> Enc -> a -> Expectation
```
Comment (lehins) [inline libs/cardano-ledger-core/testlib/Test/Cardano/Ledger/Core/Binary/Golden.hs]:
```suggestion
  (ToExpr a, DecCBOR (Annotator a), Eq a, HasCallStack) => Version -> Enc -> a -> Expectation
```

File: eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/Binary/Golden.hs [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+        it "redeemers" $ expectSuccessOnEmptyFieldRaw 5
+        -- Fields 6 and 7 should not deserialize, but they do due to a bug in the Alonzo decoder
+        -- This should not be a problem starting with PV9, because we won't allow empty lists
+        -- from there onwards
+        it "plutusV2Script" $ expectSuccessOnEmptyFieldRaw 6
+        it "plutusV3Script" $ expectSuccessOnEmptyFieldRaw 7
+        it "8th field" . expectFailureOnEmptyField 8 $
+          DecoderErrorDeserialiseFailure
+            (Binary.label $ Proxy @(Annotator (TxWits era)))
+            (DeserialiseFailure 2 "An error occured while decoding (Int,Void) not a valid key:.\nError: 8")
```
Comment (lehins) [inline eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/Binary/Golden.hs]: Let's turn this one into a property. Could I ask you to also fix a spelling mistake in "occurred"
```suggestion
        prop "Invalid field" $ \(NonNegative n) ->
          let invalidTag = n + 8
          in expectFailureOnEmptyField invalidTag $
               DecoderErrorDeserialiseFailure
                (Binary.label $ Proxy @(Annotator (TxWits era)))
                (DeserialiseFailure 2 $ "An error occurred while decoding (Int,Void) not a valid key:.\nError: " <> show invalidTag)
```

File: eras/alonzo/test-suite/test/Tests.hs [hunk 5]
```diff
@@ -29,7 +28,6 @@ defaultTests :: Spec
 defaultTests =
   describe "Alonzo tests" $ do
     AdaPreservation.tests @AlonzoEra 50
-    Golden.tests
```
Comment (lehins) [inline eras/alonzo/test-suite/test/Tests.hs:32]: Did you intend to remove these tests?

I suspect what happened was you introduced a module with the same name into `cardano-ledger-alonzo:testlib:Test.Cardano.Ledger.Alonzo.Golden` as the one in `cardano-ledger-alonzo-test:library:Test.Cardano.Ledger.Alonzo.Golden` and this resulted in cabal complaining, which led to you removing this line.
Problem is that it effectively disables those tests, which we would like to keep, regardless how ugly they look.

An easy solution would be to simply moves the contents of this file: `eras/alonzo/test-suite/src/Test/Cardano/Ledger/Alonzo/Golden.hs` intto 
`eras/alonzo/impl/test/Test/Cardano/Ledger/Alonzo/GoldenSpec.hs`.  Note that you'll also need to move the actual golden files used by that module.

File: eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/Binary/Golden.hs [hunk 6]
```diff
@@ -85,3 +132,19 @@ goldenNewEpochStateExpectation
...
+      mkBasicTxBody @ShelleyEra @TopTx
+        & certsTxBodyL .~ SSeq.fromList [testCert, testCert]
+        & ttlTxBodyL .~ SlotNo 300
+
+spec :: Spec
+spec =
+  describe "Golden" $ do
+    prop "NewEpochState" $ goldenNewEpochStateExpectation @ShelleyEra
+    describe "TxCerts" $ do
+      forEachEraVersion @ShelleyEra shelleyDecodeDuplicateDelegCertSucceeds
```
Comment (lehins) [inline eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/Binary/Golden.hs]: This needs to go into the test suite into `Test.Cardano.Ledger.Shelley.Binary.GoldenSpec`. We should only define `Spec`s that are parametric (i.e. reusable) in testlib

File: eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/Binary/Golden.hs [hunk 7]
```diff
@@ -85,3 +132,19 @@ goldenNewEpochStateExpectation
...
           ]
+
+shelleyDecodeDuplicateDelegCertSucceeds :: Version -> Spec
+shelleyDecodeDuplicateDelegCertSucceeds version =
+  it "Decodes duplicate delegation certificates successfully" $ do
+    let testCert = DelegStakeTxCert @ShelleyEra (KeyHashObj $ mkKeyHash 0) (mkKeyHash 1)
+    expectDecoderSuccessAnnWith eqRaw version (duplicateDelegCertsTxBody @ShelleyEra version) $
+      mkBasicTxBody @ShelleyEra @TopTx
+        & certsTxBodyL .~ SSeq.fromList [testCert, testCert]
+        & ttlTxBodyL .~ SlotNo 300
```
Comment (lehins) [inline eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/Binary/Golden.hs:142]: Why did you choose to add `ttl` field? It does't seem to be necessary for testing the duplicates. If you didn't add it then you'd be able to create a reusable test for shelley onwards, instead of allegra onwards like you have in this PR.

I'd recommend moving the allegra definition here and removing ttl/vldt field from the golden test

File: libs/cardano-ledger-core/testlib/Test/Cardano/Ledger/Core/Binary/Golden.hs [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+      expectationFailure $
+        "Expected a failure, but decoder succeeded:\n"
+          <> showExpr x
+
+expectDecoderResultOn ::
+  forall a b.
+  (ToExpr b, DecCBOR (Annotator a), Eq b, HasCallStack) =>
+  Version -> Enc -> a -> (a -> b) -> Expectation
+expectDecoderResultOn version enc expected f =
+  embedTripAnnExpectation version version (\x _ -> f x `shouldBeExpr` f expected) enc
```
Comment (lehins) [inline libs/cardano-ledger-core/testlib/Test/Cardano/Ledger/Core/Binary/Golden.hs]: That's the only thing that is preventing you from moving the whole module to `cardano-ledger-binary`
```suggestion
  (ToExpr b, DecCBOR (Annotator a), Eq b, HasCallStack) =>
  Version -> Enc -> a -> (a -> b) -> Expectation
expectDecoderResultOn version enc expected f =
  embedTripAnnExpectation version version (\x _ -> expectExprEqualWithMessage "" (f x) (f expected)) enc
```

File: libs/cardano-ledger-core/testlib/Test/Cardano/Ledger/Core/Binary/Golden.hs [hunk 9]
```diff
@@ -0,0 +1,81 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+
+module Test.Cardano.Ledger.Core.Binary.Golden (
```
Comment (lehins) [inline libs/cardano-ledger-core/testlib/Test/Cardano/Ledger/Core/Binary/Golden.hs]: This module should go into `cardano-ledger-binary:testlib`.

There is nothing in here that requires `cardano-ledger-core`

File: libs/cardano-ledger-core/testlib/Test/Cardano/Ledger/Core/Binary/Golden.hs [hunk 10]
```diff
@@ -0,0 +1,81 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE RankNTypes #-}
```
Comment (lehins) [inline libs/cardano-ledger-core/testlib/Test/Cardano/Ledger/Core/Binary/Golden.hs]: It doesn't look like this extension is being used
```suggestion
```

File: eras/dijkstra/impl/testlib/Test/Cardano/Ledger/Dijkstra/Binary/Golden.hs [hunk 11]
```diff
@@ -9,66 +9,117 @@
...
+      it "plutusV2Script" . expectFailureOnEmptyField 6 $
+        DecoderErrorDeserialiseFailure
+          "Annotator (MemoBytes (AlonzoTxWitsRaw DijkstraEra))"
+          (DeserialiseFailure 4 "Empty list of scripts is not allowed")
+      it "plutusV3Script" . expectFailureOnEmptyField 7 $
+        DecoderErrorDeserialiseFailure
+          "Annotator (MemoBytes (AlonzoTxWitsRaw DijkstraEra))"
+          (DeserialiseFailure 4 "Empty list of scripts is not allowed")
+      -- TODO replace this with `plutusV4Script` once that is added
+      it "8th field" . expectFailureOnEmptyField 8 $
```
Comment (lehins) [inline eras/dijkstra/impl/testlib/Test/Cardano/Ledger/Dijkstra/Binary/Golden.hs]: Same suggestion about changing this to a property test. In the matter of fact, to avoid duplication it should be defined once and reused in all eras

File: eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Binary/Golden.hs [hunk 12]
```diff
@@ -101,3 +72,103 @@ goldenListRedeemers =
       listRedeemersEnc
       (Redeemers $ Map.singleton (SpendingPurpose $ AsIx 10) (Data $ I 20, ExUnits 30 40))
       unRedeemers
+
+witsEmptyFieldWithTag :: Int -> Enc
+witsEmptyFieldWithTag k =
```
Comment (lehins) [inline eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Binary/Golden.hs]: I think it will make it clearer which tag we have in mind here
```suggestion
witsEmptyFieldWithSetTag :: Int -> Enc
witsEmptyFieldWithSetTag k =
```

File: eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Binary/Golden.hs [hunk 13]
```diff
@@ -101,3 +72,103 @@ goldenListRedeemers =
...
+
+spec :: forall era. ConwayEraTest era => Spec
+spec = do
+  describe "TxWits" $
+    describe "Empty fields not allowed" $ do
+      forEachEraVersion @era $ \version -> do
+        describe "Untagged" $ do
+          let
+            expectFailureOnEmptyField k =
+              expectDecoderFailureAnn @(TxWits era) version (witsEmptyField k)
```
Comment (lehins) [inline eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Binary/Golden.hs]: Too much copy pasting:
```suggestion
          let
            expectFailureOnEmptyField k failure = do
              let decoderError n = 
                    DecoderErrorDeserialiseFailure
                   (Binary.label $ Proxy @(Annotator (TxWits era)))
                   (DeserialiseFailure n msg)
              expectDecoderFailureAnn @(TxWits era) version (witsEmptyField k) (decoderError 4)
              expectDecoderFailureAnn @(TxWits era) version (witsEmptyFieldWithTag k) (decoderError 7)
```

File: eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Binary/Golden.hs [hunk 14]
```diff
@@ -101,3 +72,103 @@ goldenListRedeemers =
...
+              (Binary.label $ Proxy @(Annotator (TxWits era)))
+              (DeserialiseFailure 7 "Empty list found, expected non-empty")
+          it "plutusV2Script" . expectFailureOnEmptyField 6 $
+            DecoderErrorDeserialiseFailure
+              (Binary.label $ Proxy @(Annotator (TxWits era)))
+              (DeserialiseFailure 7 "Empty list of scripts is not allowed")
+          it "plutusV3Script" . expectFailureOnEmptyField 7 $
+            DecoderErrorDeserialiseFailure
+              (Binary.label $ Proxy @(Annotator (TxWits era)))
+              (DeserialiseFailure 7 "Empty list of scripts is not allowed")
```
Comment (lehins) [inline eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Binary/Golden.hs]: "If you find yourself copy pasting large chunks of code, it's likely you doing something wrong" - this was a priceless advice I learned from a professor at the university.
This whole thing is unnecessarily duplicated:
```suggestion
```

File: eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Binary/Golden.hs [hunk 15]
```diff
@@ -101,3 +72,103 @@ goldenListRedeemers =
...
+    describe "Empty fields not allowed" $ do
+      forEachEraVersion @era $ \version -> do
+        describe "Untagged" $ do
+          let
+            expectFailureOnEmptyField k =
+              expectDecoderFailureAnn @(TxWits era) version (witsEmptyField k)
+          it "addrTxWits" . expectFailureOnEmptyField 0 $
+            DecoderErrorDeserialiseFailure
+              (Binary.label $ Proxy @(Annotator (TxWits era)))
+              (DeserialiseFailure 4 "Empty list found, expected non-empty")
```
Comment (lehins) [inline eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Binary/Golden.hs]: With my suggestion above all these become:
```suggestion
          it "addrTxWits" $ expectFailureOnEmptyField 0 "Empty list found, expected non-empty"
```

**Review by lehins — CHANGES_REQUESTED** [review]
I really like the direction where we are going with these golden tests!!!
Thank you for doing this work.

File: eras/alonzo/test-suite/test/Tests.hs (hunk 5 above)
Comment (Soupstraw) [reply eras/alonzo/test-suite/test/Tests.hs:32]: I think I meant to replace it with `Test.Cardano.Ledger.Binary.Golden` and then call the old test tree from that spec, but then forgot to actually add the test here. I think merging those two modules together like you suggested makes much more sense anyways.

File: eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/Binary/Golden.hs (hunk 7 above)
Comment (Soupstraw) [reply eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/Binary/Golden.hs:142]: The TTL field is mandatory in Shelley transactions from what I can see in the CDDL.

File: eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Binary/Golden.hs (hunk 14 above)
Comment (Soupstraw) [reply eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Binary/Golden.hs]: There are slight differences in the error messages, that's why I wrote everything out like that. I'm thinking that maybe the real solution would be to ensure that the error messages are the same regardless of the code path?

File: eras/alonzo/test-suite/test/Tests.hs (hunk 5 above)
Comment (Soupstraw) [reply eras/alonzo/test-suite/test/Tests.hs:32]: I tried moving the tests out of Alonzo's `test-suite`, but it seems like it would entail moving some helper functions out of Mary's `test-suite` as well. I think this would be better done in a separate PR since there would be a lot of changes.

File: eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/Binary/Golden.hs (hunk 7 above)
Comment (Soupstraw) [reply eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/Binary/Golden.hs:142]: I could get rid of this setter, but that would mean I'd have to set the TTL field in the expected value to whatever arbitrary value is hard-coded in `mkBasicTxBody`. Then I would also have to have a different expected value for Allegra onwards, where the validity interval is not set by `mkBasicTxBody`. Basically I can re-use `shelleyDecodeDuplicateDelegCertSucceeds`, but I can no longer re-use the expected value, so it's not much of an improvement improvement in my eyes.

File: eras/alonzo/test-suite/test/Tests.hs (hunk 5 above)
Comment (Soupstraw) [reply eras/alonzo/test-suite/test/Tests.hs:32]: related https://github.com/IntersectMBO/cardano-ledger/issues/3623

File: eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/Binary/Golden.hs (hunk 7 above)
Comment (lehins) [reply eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/Binary/Golden.hs:142]: Holy cow. I never knew that ttl was mandatory in Shelley. I wonder who thought that was a good idea :sweat_smile: 

In this case we can discard my comment completely and what you did here is perfect!

File: eras/alonzo/test-suite/test/Tests.hs (hunk 5 above)
Comment (lehins) [reply eras/alonzo/test-suite/test/Tests.hs:32]: I understand your sentiment. However, I do not agree that it justifies disabling of tests. If you'd prefer not to do this in this PR, that is totally fine. In this case I would suggest you rename the `Test.Cardano.Ledger.Alonzo.Golden` module in `cardano-ledger-alonzo-test:library` to something else, so it doesn't clash with the module with same name in `cardano-ledger-alonzo:testlib`. This will at least let you re-enable these test. Just make sure you bump up the major version of `cardano-ledger-alonzo-test`, since that would be a breaking change.

Alternatively we could try using `PackageImports` extension to deal with this, which might require less work, but it's been many years since I've used it, so I am not sure what kind of pitfalls we could run into. So, I am less confident that this would be a good way to solve it.

Comment (Soupstraw) [reply eras/alonzo/test-suite/test/Tests.hs:32]: I added it back, but for some reason GitHub isn't adding an `Outdated` label to this

**Review by lehins — APPROVED** [review]
:+1:
